ballista/rust/core/proto/ballista.proto

```diff
@@ -424,54 +424,87 @@ message ExecutionGraph {
   uint64 output_partitions = 5;
   repeated PartitionLocation output_locations = 6;
   string scheduler_id = 7;
+  uint32 tid_gen = 8;
+  repeated StageAttempts failed_attempts = 9;
+}
+
+message StageAttempts {
+  uint32 stage_id = 1;
+  repeated uint32 stage_attempt_num = 2;
 }
 
 message ExecutionGraphStage {
   oneof StageType {
       UnResolvedStage unresolved_stage = 1;
       ResolvedStage resolved_stage = 2;
-      CompletedStage completed_stage = 3;
+      SuccessfulStage successful_stage = 3;
       FailedStage failed_stage = 4;
   }
 }
 
 message UnResolvedStage {
-  uint64 stage_id = 1;
-  PhysicalHashRepartition output_partitioning = 2;
-  repeated uint32 output_links = 3;
-  repeated  GraphStageInput inputs = 4;
-  bytes plan = 5;
-}
-
-message ResolvedStage {
-  uint64 stage_id = 1;
-  uint32 partitions = 2;
+  uint32 stage_id = 1;
+  uint32 stage_attempt_num = 2;
   PhysicalHashRepartition output_partitioning = 3;
   repeated uint32 output_links = 4;
   repeated  GraphStageInput inputs = 5;
   bytes plan = 6;
+  repeated string last_attempt_failure_reasons = 7;
 }
 
-message CompletedStage {
-  uint64 stage_id = 1;
-  uint32 partitions = 2;
-  PhysicalHashRepartition output_partitioning = 3;
-  repeated uint32 output_links = 4;
-  repeated  GraphStageInput inputs = 5;
-  bytes plan = 6;
-  repeated TaskStatus task_statuses = 7;
-  repeated OperatorMetricsSet stage_metrics = 8;
+message ResolvedStage {
+  uint32 stage_id = 1;
+  uint32 stage_attempt_num = 2;
+  uint32 partitions = 3;
```

Comment:
There are breaking changes here because this PR is changing the meaning of fields. For example field 3 was `PhysicalHashRepartition` and now it is `uint32`. We should assign new numbers to new fields and not modify numbers for existing fields.

Reply:
These changes meant that I could not use my benchmark client without recompiling it against this branch, which should not be necessary.

Reply:
> These changes meant that I could not use my benchmark client without recompiling it against this branch, which should not be necessary.

Well, I thought those are private structures and only used internally by the TaskManager.

Reply:
I will change the tag num back.  But still there are backward incompatible changes here,  The `stage_id` was changed from `uint64 stage_id` to `uint32 stage_id` to have the same type with other stage_ids in this proto file.

Reply:
Yes, after reviewing again, I don't think we can avoid breaking changes given the extent of these changes.